README.md

```diff
@@ -0,0 +1,206 @@
+# arm-builder-kernels
+
+## ToC
+* [Ubuntu](#ubuntu)
+* [Amazon Linux 2](#amazon-linux-2)
+
+## Ubuntu
+* update `build/kernel-src` submodule and checkout new upstream version tag
+* generate `patch/{{version-tag}}/balena.patch` from existing/previous patch(es)
+* use pull-request workflow to create a GitHub release containing Debian kernel binary (.deb)
+
+
+## Amazon Linux 2
+
+### install dependencies
+> use next generation Linux kernel as a base on a temporary AWS EC2 build instance (aarch64)
+
+    amazon-linux-extras install kernel-ng
+
+    yum groupinstall -y "Development Tools"
+
+    yum-builddep -y kernel && yum install -y pesign
+
+    yumdownloader --source kernel
+
+    groupadd mock && useradd mockbuild
+
+
+### unpack RPM
+> https://www.hiroom2.com/2016/05/29/centos-7-rebuild-kernel-with-src-rpm/
+
+    rpm -i kernel-*.src.rpm
+
+    pushd rpmbuild
+
+    cp -a SPECS/kernel.spec{,.org}
+
+
+### increment build version
+> increment build number in `SPECS/kernel.spec`
+
+    diff -uprN SPECS/kernel.spec{.org,}
+
+
+### prepare kernel source
+
+    rpmbuild -bp SPECS/kernel.spec
+
+    pushd BUILD/kernel-*
+
+    kernel_tree=$(find . -maxdepth 1 -name linux-*.amzn2.aarch64 -type d)
+
+    cp -a ${kernel_tree}{,.org}
+
+
+### patch
+> (e.g.) diff [patch](https://github.com/balena-io-archive/kernel/blob/fd12169aeb39b23b08373aa66cbd819365efec59/kernel/sys.c) [base](https://github.com/torvalds/linux/blob/v5.4/kernel/sys.c) reference
+
+```
+--- a/kernel/sys.c
++++ b/kernel/sys.c
+@@ -1196,10 +1196,25 @@ SYSCALL_DEFINE0(setsid)
+ DECLARE_RWSEM(uts_sem);
+
+ #ifdef COMPAT_UTS_MACHINE
++static char compat_uts_machine[__OLD_UTS_LEN+1] = COMPAT_UTS_MACHINE;
++
++static int __init parse_compat_uts_machine(char *arg)
++{
++   strncpy(compat_uts_machine, arg, __OLD_UTS_LEN);
++   compat_uts_machine[__OLD_UTS_LEN] = 0;
++   return 0;
++}
++early_param("compat_uts_machine", parse_compat_uts_machine);
++
++#undef COMPAT_UTS_MACHINE
++#define COMPAT_UTS_MACHINE compat_uts_machine
++#endif
++
++#ifdef COMPAT_UTS_MACHINE
+ #define override_architecture(name) \
+    (personality(current->personality) == PER_LINUX32 && \
+     copy_to_user(name->machine, COMPAT_UTS_MACHINE, \
+-             sizeof(COMPAT_UTS_MACHINE)))
++             sizeof(COMPAT_UTS_MACHINE)))
+ #else
+ #define override_architecture(name)    0
+ #endif
+@@ -1207,31 +1222,15 @@ DECLARE_RWSEM(uts_sem);
+ /*
+  * Work around broken programs that cannot handle "Linux 3.0".
+  * Instead we map 3.x to 2.6.40+x, so e.g. 3.0 would be 2.6.40
+- * And we map 4.x and later versions to 2.6.60+x, so 4.0/5.0/6.0/... would be
+- * 2.6.60.
++ * And we map 4.x to 2.6.60+x, so 4.0 would be 2.6.60.
+  */
+ static int override_release(char __user *release, size_t len)
+ {
+    int ret = 0;
+
++   strncpy(compat_uts_machine, "armv7l", __OLD_UTS_LEN);
```

Comment:
Here we override with the static string `armv7l` if `PERLINUX32` is enabled, and below, we override with `armv6l` if `UNAME26` is enabled. I suppose this has the benefit of switching between three values of `uts_machine` at runtime, but it's hacky, and doesn't allow arbitrary machine strings to be set at boot.

Reply:
@jakogut the current architecture requires x3 Docker sockets, each reporting one of these architectures:

* armv6l
* armv7l
* aarch64